controllers/model/modelversion_controller.go

```diff
@@ -0,0 +1,435 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	modelv1alpha1 "github.com/alibaba/kubedl/apis/model/v1alpha1"
+	"github.com/alibaba/kubedl/controllers/model/storage"
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// ModelVersionReconciler reconciles a ModelVersion object
+type ModelVersionReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=model.kubedl.io,resources=modelversions,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=model.kubedl.io,resources=modelversions/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=model.kubedl.io,resources=models,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=model.kubedl.io,resources=models/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",resources=persistentvolumes,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=persistentvolumeclaims,verbs=get;list;watch;create;update;patch;delete
+
+// ModelVersionController tries to create a kaniko container for each ModelVersion that triggers an image build process.
+// If a ModelVersion has its own storage backend defined, it'll also create the underlying pv and pvc first.
+func (r *ModelVersionReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	//Get the modelVersion
+	modelVersion := &modelv1alpha1.ModelVersion{}
+	err := r.Get(context.Background(), req.NamespacedName, modelVersion)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Info("modelVersion doesn't exist", "name", req.String())
+			return reconcile.Result{}, nil
+		}
+		r.Log.Error(err, "fail to get model version")
+		return reconcile.Result{}, err
+	}
+
+	// no need to build
+	if modelVersion.Status.ImageBuildPhase == modelv1alpha1.ImageBuildSucceeded ||
+		modelVersion.Status.ImageBuildPhase == modelv1alpha1.ImageBuildFailed {
+		r.Log.Info(fmt.Sprintf("image build %s", modelVersion.Status.ImageBuildPhase), "modelVersion", modelVersion.Name)
+		return reconcile.Result{}, nil
+	}
+
+	// Does the model exist
+	model := &modelv1alpha1.Model{}
+	err = r.Get(context.Background(), types.NamespacedName{Namespace: modelVersion.Namespace, Name: modelVersion.Spec.ModelName}, model)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Info("model doesn't exist", "model", modelVersion.Spec.ModelName)
+			return reconcile.Result{}, err
```

Comment:
is model is not found anymore, it should return `nil` to finish reconciling this request, otherwise it will retry endlessly.

Reply:
fixed